cosmos spark3 DataSourceV2 catalog api implementation

moderakh opened the pull request:
This PR adds support for spark3 DataSourceV2 Catalog API:
The PR is moved here. to follow or comment please use this link:
Azure#18011



```scala
spark.conf.set(s"spark.sql.catalog.cosmoscatalog", "com.azure.cosmos.spark.CosmosCatalog")
spark.conf.set(s"spark.sql.catalog.cosmoscatalog.spark.cosmos.accountEndpoint", cosmosEndpoint)
spark.conf.set(s"spark.sql.catalog.cosmoscatalog.spark.cosmos.accountKey", cosmosMasterKey)

spark.sql(s"CREATE DATABASE cosmoscatalog.mydb;")
spark.sql(s"CREATE TABLE cosmoscatalog.mydb.myContainer (word STRING, number INT) using cosmos.items 
 TBLPROPERTIES(partitionKeyPath = '/mypk', manualThroughput = '1100')")
```

Please see `CosmosCatalogSpec` for end to end integration tests.
The integration testings will work once this earlier PR merges:
https://github.com/Azure/azure-sdk-for-java/pull/17952 getting merged.

TODO: 
- There are some TODO in the code, (e.g., add support for table alter)
- the integration tests resource management needs to be figured out.
- This PR adds support for catalog metadata operation, we should also validate data operation through catalog api.

moderakh committed:
cosmos catalog api implementation

moderakh committed:
update option name

moderakh committed:
update config names

FabianMeiswinkel reviewed (approved):
Nice improvement of experience for simple operations to create table etc.

moderakh committed:
Cosmos Spark End to End Integration Test against Cosmos Emulator runs in CI (#17952)

Now we have gated CI for end to end test spark <-> cosmos db emulator.

Cosmos Spark End to End Test against Cosmos Emulator runs in CI.
- any test tagged by newly introduced tag, `RequiresCosmosEndpoint` will get included in sparkE2E test group and executed by the newly added CI, Spark_Integration_Tests_Java8 against cosmos db emulator.
- see `SparkE2EWriteSpec.scala` for end to end spark sample test. It writes data from spark to cosmos db emulator. Emulator CI 

We have two spark test groups running in the CI:
- **unit**: (only unit tests) this is the default test group.
      how to run locally in your dev machine: `mvn -e   -Dgpg.skip -Dmaven.javadoc.skip=true -Dspotbugs.skip=true -Dcheckstyle.skip=true -Drevapi.skip=true -pl ,azure-cosmos-spark_3-0_2-12 -am  clean test`
- **sparkE2E**: requires cosmos db endpoint (integration tests runs against cosmos emulator)
      how to run locally in your dev machine: `mvn -e   -Dgpg.skip -Dmaven.javadoc.skip=true -Dspotbugs.skip=true -Dcheckstyle.skip=true -Drevapi.skip=true -pl ,azure-cosmos-spark_3-0_2-12 -am  -PsparkE2E clean test`

<img width="1040" alt="Screen Shot 2020-12-03 at 9 21 28 AM" src="https://user-images.githubusercontent.com/22279672/101064552-f9374b00-3548-11eb-8ce5-63af31255864.png">

TODO:
-  we need a CI for java11.
- Cosmos Emulator requires windows, hence the current CI is running on windows, we should add a CI on Linux (targeting prod account)
- some patterns in the integration tests need to be figured out. 
-- proper resource (Database, Container) cleaning
-- proper shutdown of CosmosClient and Spark session
-- possible sharing of the CosmosClient and spark session between tests
-- which scala test style should be used?

moderakh committed:
Merge branch 'feature/cosmos/spark30' into users/moderakh/cosmos/catalog-api

moderakh commented:
retargeting this PR against Azure repo: https://github.com/Azure/azure-sdk-for-java/pull/18011
